x/stakeibc/keeper/icacallbacks_undelegate.go

```diff
@@ -81,36 +79,70 @@ func (k Keeper) UndelegateCallback(ctx sdk.Context, packet channeltypes.Packet,
 	k.Logger(ctx).Info(utils.LogICACallbackStatusWithHostZone(chainId, ICACallbackID_Undelegate,
 		icacallbackstypes.AckResponseStatus_SUCCESS, packet))
 
-	// Update delegation balances
-	err := k.UpdateDelegationBalances(ctx, hostZone, undelegateCallback)
+	// Calculate the stTokens that should be burned from the batch and get the latest
+	// completion time from the ack response
+	nativeTokensUnbonded, stTokensToBurn := k.CalculateTokensFromBatch(undelegateCallback.SplitUndelegations)
+	unbondingTime, err := k.GetLatestUnbondingCompletionTime(ctx, ackResponse.MsgResponses)
 	if err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
 
-	// Get the latest transaction completion time (to determine the unbonding time)
-	latestCompletionTime, err := k.GetLatestCompletionTime(ctx, ackResponse.MsgResponses)
+	// Update delegation balances on the validators and host zone
+	err = k.UpdateDelegationBalances(ctx, hostZone, undelegateCallback)
 	if err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
 
-	// Burn the stTokens
-	stTokenBurnAmount, err := k.UpdateHostZoneUnbondings(ctx, *latestCompletionTime, chainId, undelegateCallback)
-	if err != nil {
+	// Update the accounting on the host zone unbondings
+	if err := k.UpdateHostZoneUnbondingsAfterUndelegation(
+		ctx,
+		chainId,
+		undelegateCallback.EpochUnbondingRecordIds,
+		nativeTokensUnbonded,
+		stTokensToBurn,
+		unbondingTime,
+	); err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
-	err = k.BurnTokens(ctx, hostZone, stTokenBurnAmount)
-	if err != nil {
+
+	// Burn the stTokens from the batch
+	if err := k.BurnStTokensAfterUndelegation(ctx, hostZone, stTokensToBurn); err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
 
-	// Upon success, add host zone unbondings to the exit transfer queue
-	err = k.RecordsKeeper.SetHostZoneUnbondingStatus(ctx, chainId, undelegateCallback.EpochUnbondingRecordIds, recordstypes.HostZoneUnbonding_EXIT_TRANSFER_QUEUE)
-	if err != nil {
-		return err
+	return nil
+}
+
+// Regardless of failure/success/timeout, indicate that this ICA has completed on each validator
+// on the host zone, and on the epoch unbonding record
+func (k Keeper) MarkUndelegationAckReceived(ctx sdk.Context, hostZone types.HostZone, undelegateCallback types.UndelegateCallback) error {
+	// Indicate that this ICA has completed on each validator
+	for _, splitDelegation := range undelegateCallback.SplitUndelegations {
+		if err := k.DecrementValidatorDelegationChangesInProgress(&hostZone, splitDelegation.Validator); err != nil {
```

Comment:
To make sure I understand: this counter includes both Delegation and Undelegation changes - is that right?

Reply:
yep that's right